### ISSUE TITLE
murmurhash3 fails to install

### ISSUE THREAD

**Issue by dmontagu** [returntocorp/bento#70]
I'm unable to build murmurhash3 on MacOS (mojave; 10.14.6), which is keeping me from installing this (using python version 3.7.3).

```
pip3 install bento-cli

...

Installing collected packages: murmurhash3, cfgv, attrs, psutil, pyyaml, virtualenv, identify, aspy.yaml, nodeenv, pre-commit, docker-pycreds, websocket-client, docker, tqdm, smmap2, gitdb2, gitpython, semantic-version, bento-cli
  Running setup.py install for murmurhash3 ... error
    Complete output from command /Users/dmontague/Library/Caches/pypoetry/virtualenvs/app--3J4JXvL-py3.7/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/zs/84jkzmz96tv7krwn3tkczc6r0000gn/T/pip-install-ka4jnst1/murmurhash3/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/zs/84jkzmz96tv7krwn3tkczc6r0000gn/T/pip-record-9g12dqqk/install-record.txt --single-version-externally-managed --compile --install-headers /Users/dmontague/Library/Caches/pypoetry/virtualenvs/app--3J4JXvL-py3.7/include/site/python3.7/murmurhash3:
    running install
    running build
    running build_ext
    building 'mmh3' extension
    creating build
    creating build/temp.macosx-10.7-x86_64-3.7
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/include -arch x86_64 -I/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/include -arch x86_64 -I/Users/dmontague/Library/Caches/pypoetry/virtualenvs/app--3J4JXvL-py3.7/include -I/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/include/python3.7m -c mmh3module.cpp -o build/temp.macosx-10.7-x86_64-3.7/mmh3module.o
    warning: include path for stdlibc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
    1 warning generated.
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/include -arch x86_64 -I/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/include -arch x86_64 -I/Users/dmontague/Library/Caches/pypoetry/virtualenvs/app--3J4JXvL-py3.7/include -I/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/include/python3.7m -c murmur_hash_3.cpp -o build/temp.macosx-10.7-x86_64-3.7/murmur_hash_3.o
    warning: include path for stdlibc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
    1 warning generated.
    creating build/lib.macosx-10.7-x86_64-3.7
    g++ -bundle -undefined dynamic_lookup -L/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/lib -arch x86_64 -L/Users/dmontague/Programming/miniconda/envs/fastapi_tutorial/lib -arch x86_64 -arch x86_64 build/temp.macosx-10.7-x86_64-3.7/mmh3module.o build/temp.macosx-10.7-x86_64-3.7/murmur_hash_3.o -o build/lib.macosx-10.7-x86_64-3.7/mmh3.cpython-37m-darwin.so
    clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
    ld: library not found for -lstdc++
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    error: command 'g++' failed with exit status 1
    
    ----------------------------------------
```

I haven't looked hard into this, but based on this issue, it looks like it might not be a straightforward fix: https://github.com/hajimes/mmh3/issues/20

Let me know if there is a work around, or there is any extra information I can provide that would be helpful.

**Comment by dmontagu** [comment]
Hmm, no idea what caused this problem, but I created a new environment and it works fine. Sorry for the false alarm!

**Comment by dlukeomalley** [comment]
Reopening this issue because 2 independent users have run into this issue today.

> ... having some issues with a hashing library you guys use — the thing refuses to build and spent 30 minutes trying to get it to install with no avail. I suspect it must be a Mac Catalina thing with gcc versions (the latest build of mmh3 has Mac OS checking that fixes this). 

CC @nbrahms to prioritize this fix for today.

**Comment by dmontagu** [comment]
For what it's worth, [this issue may be related](https://stackoverflow.com/questions/52509602/cant-compile-c-program-on-a-mac-after-upgrade-to-mojave) -- I've seen a similar solution discussed with `pyenv` (https://github.com/pyenv/pyenv/wiki/Common-build-problems).

But the solution involves a relatively high-effort process of installing headers.

------

That said, looking at how mmh3 is used in the bento code, it seems to me like it doesn't really merit the extra dependency? (I recognize the potential benefits of mmh3 over cryptographic hash functions for this purpose, but I would imagine for this application it's overkill?)

This should work as a drop-in replacement for `mmh3.hash128` (which is the only use of mmh3 I see); possibly worse collision properties, but again, I feel like this is unlikely to matter in practice?:

```python
import hashlib  # builtin

hash_function = hashlib.sha512  # could change this if desired?


def hash128(input_: str) -> int:
    seed_bytes = input_.encode()
    hashed_bytes = hash_function(seed_bytes).digest()
    return int.from_bytes(hashed_bytes[:16], byteorder="big", signed=False)
```
(I literally just wrote a very similar function yesterday so was happy to share.)

(Also, sorry if I'm missing something and there are either other uses of mmh3 I'm not finding, or if it's a transitive dependency, or if you really are meaningfully benefitting from mmh3's uniformity benefits.)

**Comment by nbrahms** [comment]
Hi @dmontagu ...

Thanks for giving us a heads up about this. Would you be able to paste the output of `which gcc` and `gcc --version`, in the failing environment?

Much thanks!

**Comment by dmontagu** [comment]
It's not failing for me anymore, but I did follow the instructions in the pyenv wiki for installing the headers (linked above) right around the time I was originally facing that problem:

```
sudo installer -pkg /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg -target /
```

I'm not sure if that is related to the issue / fixes it or not, and I can't reproduce the failure any more, but presumably one of the people @dlukeomalley said are currently facing this issue might be able to confirm if it fixes it.

**Comment by dmontagu** [comment]
Also, for what it's worth, on my machine:

```
❯ which gcc
/usr/bin/gcc

~
❯ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.0 (clang-1100.0.33.12)
Target: x86_64-apple-darwin19.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
I think these are the defaults on MacOS 10.15.

**Comment by nbrahms** [comment]
Hi @dmontagu. Good to know this happens even with `clang` (previously we'd only confirmed it when a user had installed `gcc` directly).

**Comment by nbrahms** [comment]
👋  This issue should be fixed in 0.3.1b2 👋

**Comment by dlukeomalley** [comment]
Closing. 0.3.1 has officially been released!
